in_tail: Preventing incorrect inode usage from db. #8025

jinyongchoi opened the pull request:
<!-- Provide summary of changes -->

Preventing incorrect inode usage from db.

The first commit involves deleting items that are not being monitored by Fluent-Bit when it starts.

The second commit involves considering already running Fluent-Bit instances, preventing incorrect inode usage during execution.
This commit is more about enhancement option than fixing a bug. That's why 'true' and 'false' may be needed depending on the user's use case.

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->
Fixes #8025

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [x] Example configuration file for the change

- fluent-bit-test.conf

```

[SERVICE]
    flush        1
    daemon       Off
    log_level    info

    parsers_file parsers.conf
    plugins_file plugins.conf
    storage.metrics on
    storage.path /tmp/storage
    storage.sync normal
    storage.checksum off
    storage.backlog.mem_limit 5M

[INPUT]
    Name tail
    Path /tmp/testing/bulk*
    Exclude_Path *.gz,*.zip
    log_level    info
    Offset_Key   offset

    Read_from_Head true
    Refresh_Interval 1
    Rotate_Wait 1
    Inotify_Watcher false
    
    storage.type filesystem
    storage.pause_on_chunks_overlimit true

    DB /tmp/input.db
    DB.sync normal
    DB.locking false
    # new config
    DB.compare_filename true

[OUTPUT]
    name  stdout
    match *

```

- gen.sh
To account for situations where file events may not be properly detected, the creation and deletion process is repeated without using 'sleep'.
```

#!/bin/bash

max_iterations=100000

for ((iteration = 1; iteration <= max_iterations; iteration++)); do
    start_range=$((($iteration - 1) * 100 + 1))
    end_range=$(($iteration * 100))

    for i in $(seq $start_range $end_range); do
        filename="bulk$i.log"
        echo "$i." >> "$filename"
        stat -c%i "$filename"
    done

    sleep 0

    for i in $(seq $start_range $end_range); do
        filename="bulk$i.log"
        rm "$filename"
    done
done

```

- (1/2) delete stale inode from db
```
mkdir -p  /tmp/testing/
cd /tmp/testing/
rm -f /tmp/input.db
rm -f /tmp/testing/*.log

# fluentbit start
/opt/fluent-bit/bin/fluent-bit_fixed -c /opt/fluent-bit/etc/fluent-bit/fluent-bit-test.conf

# write
echo "1" > /tmp/testing/bulk.log;stat -c %i /tmp/testing/bulk.log

# fluentbit stop

# logroate and create new one.
rm -f /tmp/testing/bulk.log

# fluentbit start
/opt/fluent-bit/bin/fluent-bit_fixed -c /opt/fluent-bit/etc/fluent-bit/fluent-bit-test.conf

# log
[2023/10/19 05:50:40] [ info] [input:tail:tail.0] db: delete unmonitored stale inodes from the database: count=1

```

- (2/2) If the filename differs when retrieving the inode from the database, the inode is deleted.
```
mkdir -p  /tmp/testing/
cd /tmp/testing/
rm -f /tmp/input.db
rm -f /tmp/testing/*.log

# fluentbit start
/opt/fluent-bit/bin/fluent-bit_fixed -c /opt/fluent-bit/etc/fluent-bit/fluent-bit-test.conf

# run gen script
while true; do sh ./gen.sh; sleep 0; done

# log
# Fluentbit can read or may fail to read, but it must not read incorrectly.
[2023/10/19 06:24:14] [ info] [input:tail:tail.0] db: exists stale file from database: id=19 inode=4491280 offset=5 name=/tmp/testing/bulk126.log file_inode=4491280 file_name=/tmp/testing/bulk925.log
[2023/10/19 06:24:14] [ info] [input:tail:tail.0] db: stale file deleted from database: id=19
[2023/10/19 06:24:14] [ info] [input:tail:tail.0] db: exists stale file from database: id=20 inode=4491305 offset=5 name=/tmp/testing/bulk127.log file_inode=4491305 file_name=/tmp/testing/bulk926.log
[2023/10/19 06:24:14] [ info] [input:tail:tail.0] db: stale file deleted from database: id=20
```

- [x] Debug log output from testing the change

```
(1/2) commit
[2023/10/19 05:13:58] [ info] [input:tail:tail.0] db: delete unmonitored stale inodes from the database: count=1
```

```
(2/2) commit
[2023/10/19 05:14:03] [ info] [input:tail:tail.0] db: exists stale file from database: id=1 inode=29632163 offset=12 name=test_db.log file_inode=29632163 file_name=test_db_moved.log
[2023/10/19 05:14:03] [ info] [input:tail:tail.0] db: stale file deleted from database: id=1
```
<!--  
Please refer to the Developer Guide for instructions on building Fluent Bit with Valgrind support: 
https://github.com/fluent/fluent-bit/blob/master/DEVELOPER_GUIDE.md#valgrind
Invoke Fluent Bit and Valgrind as: $ valgrind --leak-check=full ./bin/fluent-bit <args>
-->
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found
```
valgrind --leak-check=full ./bin/flb-rt-in_tail
...
==120724== HEAP SUMMARY:
==120724==     in use at exit: 0 bytes in 0 blocks
==120724==   total heap usage: 98,919 allocs, 98,919 frees, 32,217,198 bytes allocated
==120724== 
==120724== All heap blocks were freed -- no leaks are possible
==120724== 
==120724== For lists of detected and suppressed errors, rerun with: -s
==120724== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.
- [`N/A`] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [`N/A`] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [x] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->
[#1238](https://github.com/fluent/fluent-bit-docs/pull/1238)

**Backporting**
<!--
PRs targeting the default master branch will go into the next major release usually.
If this PR should be backported to the current or earlier releases then please submit a PR for that particular branch.
-->
- [`N/A`] Backport to latest stable release.

<!--  Other release PR (not required but highly recommended for quick turnaround) -->
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

jinyongchoi commented:
This PR is now ready for review.
Thanks!

pwhelan commented on plugins/in_tail/tail_db.c:
```diff
@@ -116,11 +141,26 @@ static int db_file_exists(struct flb_tail_file *file,
+        if (ctx->compare_filename) {
+            if (flb_tail_target_file_name_cmp((char *) name, file) != 0) {
+                exists = FLB_FALSE;
+                flb_plg_info(ctx->ins, "db: exists stale file from database:"
```
This should probably be demoted to debugging.

pwhelan commented on plugins/in_tail/tail_db.c:
```diff
@@ -168,6 +208,42 @@ static int db_file_insert(struct flb_tail_file *file, struct flb_tail_config *ct
+static int stmt_add_param_concat(struct flb_tail_config *ctx,
+                                 flb_sds_t *stmt_sql, uint64_t count)
+{
+    uint64_t i;
```
please use the less ambiguous 'idx' instead of just 'i'.

pwhelan commented on plugins/in_tail/tail_db.c:
```diff
@@ -275,3 +356,121 @@ int flb_tail_db_file_delete(struct flb_tail_file *file,
+{
+    int ret = -1;
+    size_t sql_size;
+    uint64_t i;
```
same as above, for clarity's sake.

pwhelan commented on plugins/in_tail/tail_db.c:
```diff
@@ -95,16 +95,41 @@ int flb_tail_db_close(struct flb_sqldb *db)
+    int ret;
+
+    /* Bind parameters */
+    sqlite3_bind_int64(ctx->stmt_delete_file, 1, id);
```
this call should check the return value: https://www.sqlite.org/c3ref/bind_blob.html.

pwhelan commented on plugins/in_tail/tail_db.c:
```diff
@@ -116,11 +141,26 @@ static int db_file_exists(struct flb_tail_file *file,
         *id = sqlite3_column_int64(ctx->stmt_get_file, 0);
 
+        /* name: column 1 */
+        name = sqlite3_column_text(ctx->stmt_get_file, 1);
```
the return value for this call should be checked for a NULL value, just in case.

pwhelan commented on plugins/in_tail/tail_db.c:
```diff
@@ -275,3 +356,121 @@ int flb_tail_db_file_delete(struct flb_tail_file *file,
+    i = 1;
+    mk_list_foreach_safe(head, tmp, &ctx->files_static) {
+        file = mk_list_entry(head, struct flb_tail_file, _head);
+        sqlite3_bind_int64(stmt_delete_inodes, i, file->inode);
```
this call could also return an error, however unlikely it might be.

pwhelan reviewed (commented):
I made a few observations concerning error checking. The actual functionality seems to be robust and the tests seem to actually prove the patch works as expected.

edsiper commented:
@jinyongchoi thanks for your PR looking forward to merge this, would you please address the comments pointed out by @pwhelan ? thank you

jinyongchoi commented:
@pwhelan Thank you for the review.
@edsiper I have addressed the points mentioned by @pwhelan.
Please review the changes at your convenience and let me know if any further adjustments are needed. 
thank you!

RicardoAAD commented:
Tested the new implementation, and it works as expected. When the option **DB.compare_filename** is enabled, the incorrect offset issue doesn't occur

pwhelan commented:
@edsiper I have checked and all my comments have been resolved.

pwhelan reviewed (approved):
LGTM

edsiper commented:
thank you

zdyj3170101136 commented:
in k8s, this would cause re-read file after restart.

https://github.com/fluent/fluent-bit/issues/11060
